subsys/bluetooth/services/ias/ias_client.c

```diff
@@ -0,0 +1,147 @@
+static struct bt_ias_client client_list[CONFIG_BT_MAX_CONN];
+
+static struct bt_ias_client *client_by_conn(struct bt_conn *conn)
+{
+	return &client_list[bt_conn_index(conn)];
+}
+
+int bt_ias_alert_write(struct bt_conn *conn, enum bt_ias_alert_lvl alert_lvl)
+{
+	int err;
+	uint8_t lvl = (uint8_t)alert_lvl;
+
+	err = bt_gatt_write_without_response(conn,
+					client_by_conn(conn)->write.handle,
+					&lvl, sizeof(lvl), false);
+	if (err < 0) {
+		BT_ERR("IAS client alert level write failed");
+		return err;
+	}
+
+	return 0;
+}
+
+static uint8_t bt_ias_alert_lvl_disc_cb(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr,
+				struct bt_gatt_discover_params *discover)
+{
+	CHECKIF(!attr) {
+		return BT_GATT_ERR(BT_ATT_ERR_ATTRIBUTE_NOT_FOUND);
+	}
+
+	if (IS_HANDLE_VALID(attr->handle)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_HANDLE);
+	}
+
+	client_by_conn(conn)->write.handle = attr->handle;
+	return 0;
+}
+
+static uint8_t bt_ias_prim_disc_cb(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr,
+				struct bt_gatt_discover_params *discover)
+{
+	int err = 0;
+	struct bt_gatt_service_val *data;
+	struct bt_ias_client *client = client_by_conn(conn);
+
+	if (!attr) {
+		BT_DBG("Discover complete\n");
+		return BT_GATT_ITER_STOP;
+	}
+
+	data = attr->user_data;
+
+	(void)memcpy(&client->uuid, BT_UUID_ALERT_LEVEL, sizeof(client->uuid));
+	client->discover.uuid = &client->uuid.uuid;
+	client->discover.start_handle = sys_cpu_to_le16(attr->handle + 1);
+	client->discover.end_handle = sys_cpu_to_le16(data->end_handle);
+	client->discover.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+	client->discover.func = bt_ias_alert_lvl_disc_cb;
+
+	err = bt_gatt_discover(conn, &client->discover);
+
+	if (err) {
+		BT_ERR("Characteristic discover failed\n");
+		ias_client_cb->discover(conn, err);
+	}
+	ias_client_cb->discover(conn, 0);
+
+	return BT_GATT_ITER_STOP;
+}
+
+int bt_ias_discover(struct bt_conn *conn)
+{
+	int err;
+	struct bt_ias_client *client = client_by_conn(conn);
+
+	(void)memcpy(&client->uuid, BT_UUID_IAS, sizeof(client->uuid));
+	client->discover.uuid = &client->uuid.uuid;
+	client->discover.func = bt_ias_prim_disc_cb;
+	client->discover.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE;
+	client->discover.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+	client->discover.type = BT_GATT_DISCOVER_PRIMARY;
+
+	err = bt_gatt_discover(conn, &client->discover);
+
+	if (err < 0) {
+		BT_ERR("Primary discover failed(err %d\n)", err);
+		ias_client_cb->discover(conn, err);
+	}
+
+	return err;
+}
+
+int bt_ias_client_cb_register(struct bt_ias_client_cb *cb)
+{
+	CHECKIF(!cb) {
+		return -EINVAL;
+	}
+
+	CHECKIF(ias_client_cb) {
+		return -EALREADY;
+	}
+
+	ias_client_cb = cb;
+
+	return 0;
+}
+
+static void client_reset(struct bt_ias_client *ias_client)
+{
+	(void)memset(ias_client, 0, sizeof(*ias_client));
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	client_reset(client_by_conn(conn));
+}
```

Comment:
Do we need to reset the client on disconnect? Would it make more sense to reset the client when `bt_ias_discover` is called, so that we don't perform any operations on disconnecting connections that never did any IAS stuff?

Reply:
I've made a change so on discovery we are doing:

Reply:
I don't think so. 

Since we are not suing using the ias_client as a application struct, and we simply use the conn index for lookups, there's no reason to store the `bt_conn` in the ias client struct. 

A much simpler approach is to simply call `client_reset` in the `bt_ias_discover` and remove the disconnected callback, as that doesn't really serve any purpose.

Reply:
As we are doing reset in bt_ias_discover now can we mark this as resolved?